Mobile: update image size picker

enejb opened the pull request:
## Description
This PR updates the image size picker to use Bottom Sheet Control instead of the current cycle control
This helps the user pick the right size of image that they want to use. Without cycling though all of the image size and downloading all the different images. 

It also fixes an issue where we are only making the image sizes that are selectable available to be selected.

## How has this been tested?
1. Add an image block to your post. 
2. Try to switch between the different image sizes by using the size picker. 
Did it work as expected? 
Does it work the same way on the web. If you upload a smaller image then you should only see some of the sizes and not all. 

## Screenshots <!-- if applicable -->
### Before

<img src="https://user-images.githubusercontent.com/115071/118744071-7f4e0200-b808-11eb-823b-b072f94e346b.png" width="200" />

### After
<img src="https://user-images.githubusercontent.com/115071/118743708-b96ad400-b807-11eb-9850-9299a304ef70.png" width="200" />
<img src="https://user-images.githubusercontent.com/115071/118743706-b7087a00-b807-11eb-971e-7c7d78da4bd9.png" width=200 />


## Types of changes
UI update

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

enejb committed:
Update Bottom Sheet Select Control to accept a label icon

enejb committed:
Update the Image block to use Bottom Sheet Select Control

enejb committed:
Minor code fixes

enejb committed:
fix goof

enejb committed:
Fix code style

iamthomasbishop commented:
@enejb Looks good to me! Stoked we're changing Cycle into Select; makes a lot more sense!

geriux commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -113,10 +113,18 @@ export class ImageEdit extends Component {
-		} ) );
+		// Only map available image sizes.
+		this.sizeOptions = map(
+			filter( this.props.imageSizes, ( { slug } ) =>
+				get( this.props.image, [
+					'media_details',
+					'sizes',
+					slug,
+					'source_url',
+				] )
+			),
```
Hey @enejb 👋 I was testing this PR but I couldn't get the size options to show up. If I remove this filter they show up correctly. Does it work for you with the current changes?

enejb replied:
I was running into the same issue. 
I think it has to do if we know about the different sizes that are available. 

This filter was taken from https://github.com/WordPress/gutenberg/blob/7d87c9e20c15c5e1654b991f5882024fedc0fbc2/packages/block-library/src/image/image.js#L139

Currently, if you remove the filter and then try to set the size of the image that doesn't exist ( some images do not have all the different sizes) that change will not work as expected anyways. 

See https://github.com/WordPress/gutenberg/blob/7d87c9e20c15c5e1654b991f5882024fedc0fbc2/packages/block-library/src/image/edit.native.js#L65 

That being said. Maybe there is something mobile-specific that I don't quite understand when it comes to images and when we have the information about the different image sizes. 

cc: @mkevins Do you understand why sometimes the images do not have the different sizes available?

mkevins replied:
> Do you understand why sometimes the images do not have the different sizes available?

I'm not entirely sure, but I imagine that it's possible from several possibilities. One could be that the available sizes are changed at some point in the site, or possibly that the uploaded image isn't converted to the various sizes from ImageMagick. Another possibility that comes to mind, if you encountered this after uploading, is that the images sizes were not updated in the component after the upload completed?

I found this comment, which might be worth investigating: https://github.com/WordPress/gutenberg/pull/28238/files#r570481358 since the sizes come from the store, but after that performance refactor are only set in the constructor. :thinking:

mkevins replied:
As a side note, I wonder if we can remove the lodash `get` here, since optional chaining is available. Wdyt?

enejb committed:
Make sure that we always recaculate the possible sizeOptions.

Before we we calculating the sizeOptions in the constructor which didn't happen often enough. And cased issue when you first inserted the image.

enejb committed:
Remove lodash find, isEmpty, map and filter

enejb committed:
Double check that that we are dealing with arrays before proceeding.

enejb commented:
cc @geriux can you take a look at this PR again when you have a moment? 
I think it is ready for another review.

geriux commented:
> cc @geriux can you take a look at this PR again when you have a moment?
> I think it is ready for another review.

Hey @enejb 👋 I gave this another test and I saw the icon we had before is missing, was this a design requirement?

<kbd><img src="https://user-images.githubusercontent.com/4885740/124487381-ede22180-ddae-11eb-9743-a41dd2c838cd.png" width=200 /></kbd>

You can see it in the PR description screenshot:

<kbd><img src="https://user-images.githubusercontent.com/115071/118743708-b96ad400-b807-11eb-9850-9299a304ef70.png" width=200 /></kbd>

enejb committed:
Use the same icon - fullscreen

enejb commented:
Thanks for the review @geriux. It seems that I somehow must have lost the icon in a rebase. 
I updated the code so that it uses the current icon. 

See
<img width="411" alt="Screen Shot 2021-07-05 at 2 13 44 PM" src="https://user-images.githubusercontent.com/115071/124518531-7f479880-dd9b-11eb-8a8f-8773b5746846.png">

cc: @iamthomasbishop can you confirm that it is the correct icon?

geriux reviewed (approved):
LGTM! Tested it on both iOS and Android.
